docs/reactive-nonblocking.qmd

```diff
@@ -0,0 +1,316 @@
+
+## Using async/await in Shiny
+
+Asynchronous programming is a technique used in Python to increase scalability and responsiveness, usually in networked apps.
+Python supports async programming at the language level, using the `async`/`await` keywords; in the standard library, with the [`asyncio`](https://docs.python.org/3/library/asyncio.html) module; and in third-party libraries, like [FastAPI](https://fastapi.tiangolo.com/) and [aiohttp](https://docs.aiohttp.org/).
+
+Shiny has async support as well, but it's a bit different from your typical async Python framework.
+On the one hand, Shiny is built on top of [Starlette](https://www.starlette.io/), which is an async web framework, so it's possible to use async functions in many parts of your Shiny app. On the other hand, Shiny is also designed around reactive programming concepts, and that creates constraints on how async functions can be used.
+
+### Reactive async/await
+
+Shiny supports the use of `async` and `await` in your reactive code.
+You can use `async` functions in `@reactive.effect`, `@reactive.calc`, and render functions, and within those async functions you can use `await` to wait for the results of other async functions.
+
+```{shinylive-python}
+#| standalone: true
+#| components: [editor, viewer]
+import asyncio
+import datetime
+from random import random
+import time
+from shiny import App, reactive, render, ui
+
+app_ui = ui.page_fluid(
+    ui.p("The time is ", ui.output_text("current_time", inline=True)),
+    ui.hr(),
+    ui.input_numeric("x", "x", value=1),
+    ui.input_numeric("y", "y", value=2),
+    ui.input_task_button("btn", "Add numbers"),
+    ui.output_text("sum"),
+)
+
+def server(input, output, session):
+    @render.text
+    def current_time():
+        reactive.invalidate_later(1)
+        return datetime.datetime.now().strftime("%H:%M:%S %p")
+
+    @reactive.calc
+    @reactive.event(input.btn)
+    async def sum_values():
+        await asyncio.sleep(5)
+        return input.x() + input.y()
+
+    @render.text
+    async def sum():
+        return str(await sum_values())
+
+app = App(app_ui, server)
+```
+
+However, you may be surprised to learn that this technique alone does not result in improved responsiveness in Shiny apps!
+
+In the app above, the first thing in the UI is a reactive output that displays the current time.
+Click the button and notice that, during the five seconds that the app is waiting for the (artificially slow) sum calculation to complete, the time does not update.
+Despite being defined as an asynchronous function, the sum logic is still blocking the time output.
+You could replace `await asyncio.sleep(5)` with its synchronous equivalent, `time.sleep(5)`, and you'd get exactly the same result.
+
+### When async is not async
+
+While surprising, this behavior is intentional.
+Shiny goes out of its way to ensure that reactive calcs, effects, and renderers are run in a serial fashion, never concurrently.
```

Comment:
If you have a succinct way of explaining why that it is, it seems worth explaining.

```suggestion
Shiny goes out of its way to ensure that reactive functions are run in a serial fashion, never concurrently.
```